enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration will be required for OpenStack and vSphere providers
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and communicates similarly to other kubernets components.
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+A: Expect KCM perform independently to CCM and do not care about its state after initial code migration. The CCM cluster operator will be responsible for going Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+A: Certificate management should be performed by the CCM operator, like it's done in the KCM operator.
+
+Q: What happens if the KCM leader has no access to CCM? Can it continue its work? Will it give up leadership?
+A: Since KCM does not communicate with CCM it can continue to work if CCM is not available.
+
+Q: Does every node need a CCM?
+A: No. Control plane nodes only. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)
```

Comment:
All? Earlier you've stated that you will run 2 replica deployment.

Reply:
It is safe to say control-plane nodes, as they are the primary consumers for CCM and new Node creation process, and the most stable ones (workers could go away any moment). There should be some fault tolerance, so ~2 pods located on different Nodes is disired.

Reply:
All of our core components run on master nodes, always. We always run 3 pods for HA. With 2 pods you'll be having issues deciding which of the 2 is the current leader and manages CCM-related operation. Look at how kas, kcm or ks are deployed, since your deployment should match that.

Reply:
I don't think there should be any issue with running 2 if we use the leader election implementation from controller-runtime or client-go ([example](https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go)), the mechanism is just a first to grab the lock kind of thing which uses KAS as the arbiter of conflicts. As far as I'm aware this mechanism is used across OpenShift already and doesn't need 3 replicas, there's no raft style communication between different potential leaders, that's deffered to etcd.

We can make this 3 replicas if that is consistent with KAS/KCM but I don't feel like this is necessary